File: 538.convert-bst-to-greater-tree.java
/*
 * @lc app=leetcode id=538 lang=java
 *
 * [538] Convert BST to Greater Tree
 */

// @lc code=start

import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode convertBST(TreeNode root) {
        if (root == null) return null;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        int acc = 0;
        while (!stack.empty()) {
            TreeNode node = stack.pop();
            if (node != null) {
                if (node.left != null) stack.push(node.left);
                stack.push(node); stack.push(null);
                if (node.right != null) stack.push(node.right);
            } else {
                node = stack.pop();
                acc += node.val;
                node.val = acc;
            }
        }
        return root;
    }
}
// @lc code=end

File: 538.convert-bst-to-greater-tree.cpp
/*
 * @lc app=leetcode id=538 lang=cpp
 *
 * [538] Convert BST to Greater Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <stack>

using namespace std;

class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        if (root == nullptr) return nullptr;
        stack<TreeNode *> st;
        st.push(root);
        int acc = 0;
        while (!st.empty()) {
            TreeNode *node = st.top(); st.pop();
            if (node != nullptr) {
                if (node->left != nullptr) st.push(node->left);
                st.push(node); st.push(nullptr);
                if (node->right != nullptr) st.push(node->right);
            } else {
                node = st.top(); st.pop();
                node->val += acc;
                acc = node->val;
            }
        }
        return root;
    }
};
// @lc code=end
